httpcore/_backends/trio.py

```diff
@@ -55,15 +54,17 @@ async def start_tls(
 
     async def read(self, n: int, timeout: TimeoutDict) -> bytes:
         read_timeout = none_as_inf(timeout.get("read"))
-        exc_map = {trio.TooSlowError: ReadTimeout, trio.BrokenResourceError: ReadError}
+        exc_map: dict = {trio.TooSlowError: ReadTimeout}
 
         async with self.read_lock:
             with map_exceptions(exc_map):
                 with trio.fail_after(read_timeout):
-                    data = await self.stream.receive_some(max_bytes=n)
-                    if data == b"":
-                        raise ReadError("Server disconnected while attempting read")
-                    return data
+                    try:
+                        return await self.stream.receive_some(max_bytes=n)
+                    except trio.BrokenResourceError:
```

Comment:
Given the Trio docs, I think it's *possible* that we're only hitting `BrokenResourceError` here, because we're failing to handle the `b""` case first, and then continuing to read from the socket, because trio *does* return `b""` when the socket EOFs.

However it's not 100% clear to me if that's the case or if we hit this case on unclean closes on the remote end.

https://trio.readthedocs.io/en/stable/reference-io.html#trio.SocketStream.receive_some

Reply:
Short... I reckon we should fix our `h11` and `h2` handling of `b""`, but *not* add this `trio.BrokenResourceError` catch case, since it's not clear to me that it's needed/correct.